File: Python/94-binaryTreeInorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
    def __init__(self):
        self.stack = []
        self.res   = []

    def inorderTraversal(self, root):
        self.recursive_traverse(root)
        return self.res

    def recursive_traverse(self, node):
        if node:
           self.recursive_traverse(node.left)
           self.res.append(node.val)
           self.recursive_traverse(node.right)
        return

    def iterTraverse(self, root):
        while root or self.stack:
              if root:
                 self.stack.append(root)
                 root = root.left
              else:
                 root = self.stack.pop()
                 self.res.append(root.val)
                 root = root.right
        return

  

File: C++/94-BinaryTreeInorderTraversal.cpp
#include <vector>

using namespace std;
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Solution {
public:
    vector<int> inorderTraversalI(TreeNode* root) {
                vector<int> res;
                iterInorderHelper(root, res);
                return res;
    }
    vector<int> inorderTraversal(TreeNode* root) {
                vector<int> res;
                vector<TreeNode*> stack;
                while (root || !stack.empty()) {
                       if (root) {
                           stack.push_back(root);
                           root = root->left;
                       }
                       else {
                           root = stack.back();
                           res.push_back(root->val);
                           stack.pop_back();
                           root = root->right;
                       };
                };
                return res;
    }
private:
    void recInorderHelper(TreeNode* root, vector<int>& res){
         if (root != nullptr){
            recInorderHelper(root->left, res);
            res.push_back(root->val);
            recInorderHelper(root->right, res);
         };
         return;
    }
    // Chris:TODO::NTR!
};
int main(void){
    return 0;
}